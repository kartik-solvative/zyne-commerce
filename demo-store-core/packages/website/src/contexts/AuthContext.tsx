import { createContext, useContext } from 'react'

type Context = {
  accessToken: string
  domain: string
  endpoint: string
  organization: string
}

type Props = {
  children?: React.ReactNode
} & Context

const AuthContext = createContext<Partial<Context>>({})

export const AuthProvider: React.FC<Props> = ({ children, ...props }) => (
  <AuthContext.Provider value={{ ...props }}>
    {children}
  </AuthContext.Provider>
)

export const useAuthContext = () => useContext(AuthContext)
