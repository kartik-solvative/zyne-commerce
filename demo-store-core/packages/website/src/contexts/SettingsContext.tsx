import { getRawDataOrganization } from '#data/organization'
import { getLocale, Locale } from '#i18n/locale'
import { memoize } from '#utils/memoize'
import type { RawDataOrganization } from '@commercelayer/demo-store-types'
import { createContext, useContext } from 'react'

export type SettingsContext = {
  organization: RawDataOrganization & {
    slug: string | null
  }
  locale: Locale
}

const SettingsContext = createContext<Partial<SettingsContext>>({})

export const SettingsProvider: React.FC<SettingsContext & { children: React.ReactNode }> = ({ children, ...props }) => (
  <SettingsContext.Provider value={{ ...props }}>
    {children}
  </SettingsContext.Provider>
)

export const useSettingsContext = () => useContext(SettingsContext)

export const serverSideSettings = memoize(
  async (localeCode: string): Promise<SettingsContextProps> => {
    return {
      settingsContext: {
        organization: {
          ...await getRawDataOrganization(),
          slug: process.env.NEXT_PUBLIC_CL_ENDPOINT?.match(/^https?:\/\/(?<slug>[\w-]+)/)?.groups?.slug || null
        },
        locale: await getLocale(localeCode),
      }
    }
  }
)

type SettingsContextProps = {
  settingsContext: SettingsContext
}
