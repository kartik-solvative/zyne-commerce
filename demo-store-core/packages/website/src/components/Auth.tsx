import { AuthProvider } from '#contexts/AuthContext'
import { useSettingsContext } from '#contexts/SettingsContext'
import type { ShoppableLocale } from '#i18n/locale'
import { NEXT_PUBLIC_BASE_PATH } from '#utils/envs'
import { getPersistKey } from '#utils/order'
import { AuthReturnType, ClientCredentials, getSalesChannelToken } from '@commercelayer/js-auth'
import { CommerceLayer, LineItemsContainer, OrderContainer, OrderStorage } from '@commercelayer/react-components'
import type { DefaultChildrenType } from '@commercelayer/react-components/lib/esm/typings/globals'
import { useRouter } from 'next/router'
import { useEffect, useState } from 'react'

type Auth = {
  accessToken: string
  refreshToken?: string
  expires?: number
  tokenType: string
}

const getClientCredentials = (clientId: string, endpoint: string, market: number): ClientCredentials => ({
  clientId,
  endpoint,
  scope: `market:${market}`
})

const getAuth = (market: number): Auth | null => {
  const storeKey = getStoreKey(market)
  return JSON.parse(localStorage.getItem(storeKey) || 'null')
}

const storeAuth = (market: number, authReturn: Awaited<AuthReturnType>): Auth | null => {
  if (!authReturn) {
    return null
  }

  const storeKey = getStoreKey(market)

  const auth: Auth = {
    tokenType: authReturn.tokenType,
    accessToken: authReturn.accessToken,
    expires: authReturn.expires?.getTime(),
    refreshToken: authReturn.refreshToken
  }

  localStorage.setItem(storeKey, JSON.stringify(auth))

  return auth
}

const getStoreKey = <M extends number>(market: M): `clayer_token-market:${M}` => `clayer_token-market:${market}`

const hasExpired = (time: number | undefined): boolean => time === undefined || time < Date.now()

const isValid = (auth: Auth | null): auth is Auth => !hasExpired(auth?.expires)

type Props = {
  children: DefaultChildrenType
  locale: ShoppableLocale
}

export const Auth: React.FC<Props> = ({ children, locale }) => {

  const router = useRouter()
  const settings = useSettingsContext()

  const [market, setMarket] = useState<number | undefined>(settings.locale?.isShoppable ? settings.locale?.country.market : undefined)
  const [auth, setAuth] = useState<Auth | null>(null)

  const clientId = process.env.NEXT_PUBLIC_CL_CLIENT_ID
  const endpoint = process.env.NEXT_PUBLIC_CL_ENDPOINT

  useEffect(function updateMarket() {
    if (settings.locale?.isShoppable && settings.locale?.country.market !== market) {
      setMarket(settings.locale?.country.market)
    }
  }, [settings.locale, market])

  useEffect(function updateAccessToken() {
    let isMounted = true

    if (market === undefined || clientId === undefined || endpoint === undefined) {
      setAuth(null)
      return
    }

    const storedAuth: Auth | null = getAuth(market)
    const authIsValid = isValid(storedAuth)

    if (authIsValid) {
      setAuth(storedAuth)
    } else {
      getSalesChannelToken(getClientCredentials(clientId, endpoint, market))
        .then(authReturn => {
          if (isMounted) {
            setAuth(storeAuth(market, authReturn))
          }
        })
    }

    return () => {
      isMounted = false
    }
  }, [market, router.asPath, clientId, endpoint])

  if (!auth || !endpoint) {
    return (
      <>
        <CommerceLayer accessToken='' endpoint=''>
          <OrderContainer>
            <LineItemsContainer>
              {children}
            </LineItemsContainer>
          </OrderContainer>
        </CommerceLayer>
      </>
    )
  }

  const { hostname } = new URL(endpoint)
  const [, organization, domain] = hostname.match(/^(.*).(commercelayer.(co|io))$/) || []

  const return_url = typeof window !== 'undefined' ? `${window.location.protocol}//${window.location.host}${NEXT_PUBLIC_BASE_PATH}/${locale.code}` : undefined
  const cart_url = typeof window !== 'undefined' ? `${window.location.protocol}//${window.location.host}${NEXT_PUBLIC_BASE_PATH}/${locale.code}/cart` : undefined

  return (
    <AuthProvider accessToken={auth.accessToken} endpoint={endpoint} organization={organization} domain={domain}>
      <CommerceLayer accessToken={auth.accessToken} endpoint={endpoint}>
        <OrderStorage persistKey={getPersistKey(locale)}>
          <OrderContainer attributes={{
            language_code: locale.language.code,
            return_url,
            cart_url
          }}>
            <LineItemsContainer>
              {children}
            </LineItemsContainer>
          </OrderContainer>
        </OrderStorage>
      </CommerceLayer>
    </AuthProvider>
  )
}
