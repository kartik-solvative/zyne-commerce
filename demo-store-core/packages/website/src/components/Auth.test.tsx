import { useAuthContext } from '#contexts/AuthContext'
import { SettingsProvider } from '#contexts/SettingsContext'
import type { AuthReturnType as OriginalAuthReturnType } from '@commercelayer/js-auth'
import CommerceLayerContext from '../../../../node_modules/@commercelayer/react-components/lib/cjs/context/CommerceLayerContext'
import { act, render, screen } from '@testing-library/react'
import { createLocale, createOrganization, createRouter } from 'jest.helpers'
import { useContext } from 'react'
import { Auth } from './Auth'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')
const getSalesChannelToken = jest.spyOn(require('@commercelayer/js-auth'), 'getSalesChannelToken')

type AuthReturnType = Pick<NonNullable<Awaited<OriginalAuthReturnType>>, 'tokenType' | 'accessToken' | 'expires' | 'refreshToken'>

beforeEach(() => {
  useRouter.mockReset()
})

const ContextTester = () => {
  const { accessToken: clAccessToken, endpoint: clEndpoint } = useContext(CommerceLayerContext)
  const { accessToken, endpoint, organization, domain } = useAuthContext()

  return (
    <>
      <div data-testid="clAccessToken">clAccessToken: {clAccessToken}</div>
      <div data-testid="clEndpoint">clEndpoint: {clEndpoint}</div>
      <div data-testid="accessToken">accessToken: {accessToken}</div>
      <div data-testid="endpoint">endpoint: {endpoint}</div>
      <div data-testid="organization">organization: {organization}</div>
      <div data-testid="domain">domain: {domain}</div>
    </>
  )
}

test('should match the snapshot', async () => {
  process.env.NEXT_PUBLIC_CL_CLIENT_ID = 'client-1234'
  process.env.NEXT_PUBLIC_CL_ENDPOINT = 'https://demo-store.commercelayer.co'

  useRouter.mockImplementation(() => createRouter('/'))

  const authReturn: AuthReturnType = {
    tokenType: 'bearer',
    accessToken: 'accessToken-1234',
    expires: new Date(0),
    refreshToken: 'refreshToken-1234'
  }

  getSalesChannelToken.mockResolvedValue(authReturn)

  let container
  await act(async () => {
    ({ container } = render(<Auth locale={createLocale()}><ContextTester /></Auth>))
  })

  expect(container).toMatchSnapshot()
})

test('should fetch accessToken and set it properly when "locale" is set', async () => {
  process.env.NEXT_PUBLIC_CL_CLIENT_ID = 'client-1234'
  process.env.NEXT_PUBLIC_CL_ENDPOINT = 'https://demo-store.commercelayer.co'

  useRouter.mockImplementation(() => createRouter('/'))

  const authReturn: AuthReturnType = {
    tokenType: 'bearer',
    accessToken: 'accessToken-1234',
    expires: new Date(0),
    refreshToken: 'refreshToken-1234'
  }

  getSalesChannelToken.mockResolvedValue(authReturn)

  await act(async () => {
    render(
      <SettingsProvider locale={createLocale()} organization={{ ...createOrganization(), slug: 'commerce-layer' }}>
        <Auth locale={createLocale()}><ContextTester /></Auth>
      </SettingsProvider>
    )
  })

  expect(getSalesChannelToken).toBeCalledWith({
    clientId: 'client-1234',
    endpoint: 'https://demo-store.commercelayer.co',
    scope: 'market:123456789'
  })

  expect(localStorage.getItem('clayer_token-market:123456789')).toEqual(JSON.stringify({
    tokenType: 'bearer',
    accessToken: 'accessToken-1234',
    expires: 0,
    refreshToken: 'refreshToken-1234'
  }))

  expect(screen.getByTestId('clAccessToken')).toHaveTextContent('accessToken-1234')
  expect(screen.getByTestId('clEndpoint')).toHaveTextContent('https://demo-store.commercelayer.co')

  expect(screen.getByTestId('accessToken')).toHaveTextContent('accessToken-1234')
  expect(screen.getByTestId('endpoint')).toHaveTextContent('https://demo-store.commercelayer.co')
  expect(screen.getByTestId('organization')).toHaveTextContent('demo-store')
  expect(screen.getByTestId('domain')).toHaveTextContent('commercelayer.co')

})
