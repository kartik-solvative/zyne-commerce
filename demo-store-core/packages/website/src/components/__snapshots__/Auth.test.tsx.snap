// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div>
  <div
    data-testid="clAccessToken"
  >
    clAccessToken: 
  </div>
  <div
    data-testid="clEndpoint"
  >
    clEndpoint: 
  </div>
  <div
    data-testid="accessToken"
  >
    accessToken: 
  </div>
  <div
    data-testid="endpoint"
  >
    endpoint: 
  </div>
  <div
    data-testid="organization"
  >
    organization: 
  </div>
  <div
    data-testid="domain"
  >
    domain: 
  </div>
</div>
`;
